### ISSUE TITLE
Input Causes Panic: {:{

### ISSUE THREAD

**Issue by BrianHicks** [hashicorp/hcl#128]
I've found a crashing input in HCL as of d7400db7143f8e869812e50a53acd6c8d92af3b8. Here's an [SSCCE](http://www.sscce.org/) demonstrating the problem.

```go
package main

import (
	"github.com/hashicorp/hcl"
)

var (
	data = 	"{:{"
)

func main() {
	var out interface{}
	hcl.Unmarshal([]byte(data), &out)
}
```

When run, this gives the following output:

```
panic: runtime error: index out of range

goroutine 1 [running]:
panic(0x1ced60, 0xc82000a0b0)
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/runtime/panic.go:481 +0x3e6
github.com/hashicorp/hcl/hcl/ast.(*ObjectList).Pos(0xc820010200, 0x0, 0xc8200413b0, 0xc8200413e0, 0xc8200413b0, 0x15)
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/github.com/hashicorp/hcl/hcl/ast/ast.go:112 +0x11f
github.com/hashicorp/hcl.(*decoder).decodeMap(0xc820041d70, 0x1fc790, 0x4, 0x383058, 0xc820010200, 0x19b160, 0xc820012150, 0x15, 0x0, 0x0)
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/github.com/hashicorp/hcl/decoder.go:343 +0x1083
github.com/hashicorp/hcl.(*decoder).decode(0xc820041d70, 0x1fc790, 0x4, 0x383058, 0xc820010200, 0x19a5a0, 0xc82000a120, 0x194, 0x0, 0x0)
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/github.com/hashicorp/hcl/decoder.go:100 +0xcc6
github.com/hashicorp/hcl.(*decoder).decodeInterface(0xc820041d70, 0x1fc790, 0x4, 0x383058, 0xc820010200, 0x19a5a0, 0xc82000a120, 0x194, 0x0, 0x0)
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/github.com/hashicorp/hcl/decoder.go:284 +0x6de
github.com/hashicorp/hcl.(*decoder).decode(0xc820041d70, 0x1fc790, 0x4, 0x383058, 0xc820010200, 0x19a5a0, 0xc82000a120, 0x194, 0x0, 0x0)
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/github.com/hashicorp/hcl/decoder.go:98 +0xc14
github.com/hashicorp/hcl.DecodeObject(0x18d2e0, 0xc82000a120, 0x383058, 0xc820010200, 0x0, 0x0)
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/github.com/hashicorp/hcl/decoder.go:60 +0x409
github.com/hashicorp/hcl.Unmarshal(0x413000, 0x3, 0x200000, 0x18d2e0, 0xc82000a120, 0x0, 0x0)
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/github.com/hashicorp/hcl/decoder.go:32 +0x192
github.com/BrianHicks/fuzz-hcl.ParseSafely(0x413000, 0x3, 0x200000, 0x0, 0x0, 0x0, 0x0)
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/github.com/BrianHicks/fuzz-hcl/fuzz.go:20 +0xc1
github.com/BrianHicks/fuzz-hcl.Fuzz(0x413000, 0x3, 0x200000, 0xc820070058)
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/github.com/BrianHicks/fuzz-hcl/fuzz.go:11 +0x105
go-fuzz-dep.Main(0x23a3e8)
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/go-fuzz-dep/main.go:49 +0x14c
main.main()
	/var/folders/f6/l26tk3xs5pjczcd5k5pcrmk00000gn/T/go-fuzz-build846495901/src/go.fuzz.main/main.go:10 +0x23
exit status 2
```

(and here it is when run through panicparse, for convenience)

```
panic: runtime error: index out of range

exit status 2
To see all goroutines, visit https://github.com/maruel/panicparse#GOTRACEBACK

1: running
                panic.go:481   panic(0x1ced60, 0xc82000a0b0)
    ast         ast.go:112     (*ObjectList).Pos(#2, 0, #3, 0xc8200413e0, #3, 0x15)
    hcl         decoder.go:343 (*decoder).decodeMap(#4, 0x1fc790, 0x4, 0x383058, #2, 0x19b160, 0xc820012150, 0x15, 0, 0)
    hcl         decoder.go:100 (*decoder).decode(#4, 0x1fc790, 0x4, 0x383058, #2, 0x19a5a0, #1, 0x194, 0, 0)
    hcl         decoder.go:284 (*decoder).decodeInterface(#4, 0x1fc790, 0x4, 0x383058, #2, 0x19a5a0, #1, 0x194, 0, 0)
    hcl         decoder.go:98  (*decoder).decode(#4, 0x1fc790, 0x4, 0x383058, #2, 0x19a5a0, #1, 0x194, 0, 0)
    hcl         decoder.go:60  DecodeObject(0x18d2e0, #1, 0x383058, #2, 0, 0)
    hcl         decoder.go:32  Unmarshal(0x413000, 0x3, 0x200000, 0x18d2e0, #1, 0, 0)
    fuzz-hcl    fuzz.go:20     ParseSafely(0x413000, 0x3, 0x200000, 0, 0, 0, 0)
    fuzz-hcl    fuzz.go:11     Fuzz(0x413000, 0x3, 0x200000, 0xc820070058)
    go-fuzz-dep main.go:49     Main(0x23a3e8)
    main        main.go:10     main()
```

**Comment by BrianHicks** [comment]
More inputs that produce the same panic message:

```
	"t t e{{}}"
	"{:{"
	"o{{}}"
	"t t e d N{{}}"
	"t t e d{{}}"
	"t=0t d{}"
	"N{}N{{}}"
	"v\nN{{}}"
	"v=/\n[,"
	"v=0E0v d{}"
```

**Comment by mitchellh** [comment]
Fixed the initial comment here. Looking into others just to verify.

**Comment by mitchellh** [comment]
All fixes for inputs here in #134
